### PR TITLE
Protobuf definitions and DB schema

### PR DESCRIPTION
**Pull request by arjunsingh1212**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -15,8 +19,114 @@ repositories {
...
+
+    // https://mvnrepository.com/artifact/com.google.protobuf/protobuf-java
+	compile group: 'com.google.protobuf', name: 'protobuf-java', version: '3.11.1'
+	compile group: 'com.google.protobuf', name: 'protobuf-java-util', version: '3.11.1'
+
+	// https://mvnrepository.com/artifact/io.grpc/grpc-all
+	compile group: 'io.grpc', name: 'grpc-all', version: '1.26.0'
+
+	// https://mvnrepository.com/artifact/io.grpc/grpc-services
+	compile group: 'io.grpc', name: 'grpc-services', version: '1.28.0'
```
Comment (thejasn) [inline build.gradle]: get the latest deps

File: src/main/proto/catalogue.proto [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+  //Add an user to DB
+  rpc CreateUser(CreateUserRequest) returns (User);
+
+  //Delete a specific user
+  rpc DeleteUser(DeleteUserRequest) returns (google.protobuf.Empty);
+
+  //Get a specific user
+  rpc GetUser(GetUserRequest) returns (User);
+  //Get all Users
+  rpc GetUsers(GetUsersRequest) returns (stream Users);
```
Comment (thejasn) [inline src/main/proto/catalogue.proto]:
```suggestion
  rpc StreamUser(GetUsersRequest) returns (stream Users);
```

Comment (thejasn) [inline src/main/proto/catalogue.proto]:
```suggestion
  rpc GetUsers(GetUsersRequest) returns (StreamUsersResponse);
```

File: src/main/proto/catalogue.proto [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+  rpc DeleteItem(DeleteItemRequest) returns (google.protobuf.Empty);
+
+  //Get an specific Item
+  rpc GetItem(GetItemRequest) returns (Item);
+  //Get all items
+  rpc GetItems(GetItemsRequest) returns (stream Items);
+  //Get all items of a specific catalogue
+  rpc GetItemsCatalogue(GetItemsCatalogueRequest) returns (stream ItemsCatalogue);
+  //Get all items of a specific user
+  rpc GetItemsUser(GetItemsUserRequest) returns (stream ItemsUser);
```
Comment (thejasn) [inline src/main/proto/catalogue.proto]: use a oneof in `GetItemsRequest` for the other 2 rpcs as well

File: src/main/proto/catalogue.proto [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+
+
+//Request to get all users
+message GetUsersRequest {
+}
+
+//User Data
+message User {
+  //Id of user
+  int32 user_id = 1;
```
Comment (thejasn) [inline src/main/proto/catalogue.proto]: make it string, and use UUID

File: src/main/proto/catalogue.proto [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+  //List of users (To be used in stream)
+  User user = 1;
+}
+
+//Catalogue Data
+message Catalogue {
+  //Id of catalogue
+  int32 catalogue_id = 1;
+  //Id of user whom the catalogue belongs
+  int32 user_id = 2;
```
Comment (thejasn) [inline src/main/proto/catalogue.proto]: revisit fields (missing)

File: src/main/proto/catalogue.proto [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+message Catalogues {
+  //List of catalogues (To be used in stream)
+  Catalogue catalogue = 1;
+}
+
+//More than one catalogues of specific user
+message CataloguesUser {
+  //List of users (To be used in stream)
+  Catalogue catalogue = 1;
+}
```
Comment (thejasn) [inline src/main/proto/catalogue.proto]: move to oneof

File: src/main/proto/catalogue.proto [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+  //Id of item
+  int32 item_id = 1;
+  //If of catalogue in which the item belongs
+  int32 catalogue_id = 2;
+  //Name of item
+  string name = 3;
+  //Price of item
+  int64 price = 4;
+  //Quantity or number of instances of the item
+  int32 quantity = 5;
```
Comment (thejasn) [inline src/main/proto/catalogue.proto]: doube and unsigned int

File: src/main/proto/catalogue.proto [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+//More than one item of some specific user
+message ItemsUser {
+  //List of items (To be used in stream)
+  Item item = 1;
+}
+
+//Data for item creation request
+message CreateItemRequest {
+  // The parent resource
+  string parent = 1;
```
Comment (thejasn) [inline src/main/proto/catalogue.proto]: remove parent

File: src/main/proto/catalogue.proto [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+message DeleteItemRequest {
+  //Sending name of item
+  string name = 1;
+}
+
+//Request to get an item
+message GetItemRequest {
+  //Sending name of item
+  string name = 1;
+}
```
Comment (thejasn) [inline src/main/proto/catalogue.proto]: use id

File: src/main/resources/DatabaseSchema.sql [hunk 10]
```diff
@@ -0,0 +1,108 @@
+-- MySQL dump 10.13  Distrib 8.0.11, for macos10.13 (x86_64)
+--
+-- Host: localhost    Database: assignment4
+-- ------------------------------------------------------
+-- Server version	8.0.25
+
+/*!40101 SET @OLD_CHARACTER_SET_CLIENT=@@CHARACTER_SET_CLIENT */;
+/*!40101 SET @OLD_CHARACTER_SET_RESULTS=@@CHARACTER_SET_RESULTS */;
+/*!40101 SET @OLD_COLLATION_CONNECTION=@@COLLATION_CONNECTION */;
```
Comment (thejasn) [inline src/main/resources/DatabaseSchema.sql:9]: use snake case

Comment (thejasn) [reply src/main/resources/DatabaseSchema.sql:9]: add metadata `created_at` and `updated_at`

File: src/main/proto/catalogue.proto [hunk 11]
```diff
@@ -0,0 +1,244 @@
+syntax="proto3";
+
```
Comment (thejasn) [inline src/main/proto/catalogue.proto:4]: add package name

File: build.gradle (hunk 1 above)
Comment (arjunsingh1212) [reply build.gradle]: Done. Moved to v1.38.0

File: src/main/proto/catalogue.proto (hunk 2 above)
Comment (arjunsingh1212) [reply src/main/proto/catalogue.proto]: Done the changes

Comment (arjunsingh1212) [reply src/main/proto/catalogue.proto]: Done the changes

File: src/main/proto/catalogue.proto (hunk 3 above)
Comment (arjunsingh1212) [reply src/main/proto/catalogue.proto]: Used oneof feature

File: src/main/proto/catalogue.proto (hunk 4 above)
Comment (arjunsingh1212) [reply src/main/proto/catalogue.proto]: Done this change

File: src/main/proto/catalogue.proto (hunk 5 above)
Comment (arjunsingh1212) [reply src/main/proto/catalogue.proto]: Added more fields

File: src/main/proto/catalogue.proto (hunk 6 above)
Comment (arjunsingh1212) [reply src/main/proto/catalogue.proto]: Done

File: src/main/proto/catalogue.proto (hunk 7 above)
Comment (arjunsingh1212) [reply src/main/proto/catalogue.proto]: Done.

File: src/main/proto/catalogue.proto (hunk 8 above)
Comment (arjunsingh1212) [reply src/main/proto/catalogue.proto]: Removed

File: src/main/proto/catalogue.proto (hunk 9 above)
Comment (arjunsingh1212) [reply src/main/proto/catalogue.proto]: Done the changes

File: src/main/resources/DatabaseSchema.sql (hunk 10 above)
Comment (arjunsingh1212) [reply src/main/resources/DatabaseSchema.sql:9]: Created the new fields of creation time and updation time. And used snake case.

File: src/main/proto/catalogue.proto (hunk 11 above)
Comment (arjunsingh1212) [reply src/main/proto/catalogue.proto:4]: Added

**Review by thejasn — APPROVED** [review]

**Pull request merged**